aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -252,16 +252,19 @@ SparseTensor& copy_sparse_(SparseTensor& self, const SparseTensor& src) {
   return self;
 }
 
+// NOTE: `coalesce` should never be an in-place operation.
 SparseTensor coalesce_sparse_cpu(const SparseTensor& self) {
   AT_ASSERT(self.defined());
   AT_ASSERT(!self.is_variable());
   AT_ASSERT(self.is_sparse());
 
   if (self._nnz() < 2) {
-    _get_sparse_impl(self)->set_coalesced(true);
+    SparseTensor dst = self.clone();
```

Comment:
Does this have to be a clone? Why can't it behave like contiguous and return a view-like new tensor when it is coalesced?

Reply:
The weirdness here is that coalesce is a not-always-accurate cache, i.e. if you create a new sparse tensor with nnz < 2, it will report is_coalesced as False.  We don't want to change the data structure out from under someone if they do all out-of-place-ops, e.g.:
```
x = # create sparse tensor
assertFalse(x.is_coalesced)
y = x.coalesce()
assertFalse(x.is_coalesced)
```

So, I think the right thing to do here, is move the `is_coalesced` case up and don't clone in that case; then, the other cases should clone and not change the underlying data structure.